### PR TITLE
statistics: separte Get/Put according to query for statsCacheInner

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/34052

Problem Summary:

### What is changed and how it works?

separte Get/Put according to query for statsCacheInner in order to let only user query element affected by lru later

The get/put will used put/get by query in following cases:

1. sync loads in optimizer
2. get table stats in query
3. manually analyze table

### Check List

Tests 

- [x] Unit test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by xuyifangreeneyes — APPROVED** [review]

File: statistics/handle/handle_hist.go [hunk 1]
```diff
@@ -191,7 +191,7 @@ func (h *Handle) HandleOneTask(lastTask *NeededColumnTask, readerCtx *StatsReade
 	}
 	col := task.TableColumnID
 	oldCache := h.statsCache.Load().(statsCache)
-	tbl, ok := oldCache.Get(col.TableID)
+	tbl, ok := oldCache.GetByQuery(col.TableID)
```
Comment (chrysan) [inline statistics/handle/handle_hist.go:194]: I think it's duplicated with GetTableStats.

**Review by chrysan — COMMENTED** [review]
LGTM

File: executor/analyze.go [hunk 2]
```diff
@@ -242,7 +243,7 @@ func (e *AnalyzeExec) Next(ctx context.Context, req *chunk.Chunk) error {
 	if err != nil {
 		e.ctx.GetSessionVars().StmtCtx.AppendWarning(err)
 	}
-	return statsHandle.Update(e.ctx.GetInfoSchema().(infoschema.InfoSchema))
+	return statsHandle.Update(e.ctx.GetInfoSchema().(infoschema.InfoSchema), handle.WithTableStatsByQuery())
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply executor/analyze.go:249]: nice cache, updated.

File: statistics/handle/handle_hist.go (hunk 1 above)
Comment (Yisaer) [reply statistics/handle/handle_hist.go]: updated.

**Pull request merged**